### PR TITLE
Add support for function pointers with defaultable parameters

### LINKED ISSUE #36020: Closure for default parameters
Assignee: ChiranS
Points: 4

### LINKED ISSUE #37123: `tSymbol` is not updated properly for function type parameters in the lang.array module
**Description:**
There are several methods in the ballerina/lang.array module which requires a function as a parameter. Those functions are `'map`, `forEach`, `reduce`, `'filter`, `some`, `every`. 

When creating the `symbols` for each method inside the `BIRPackageSymbolEnter` class for there are two parameters. One is an array and the second one is a function. For the function type parameter there is a return type and one more parameters.  This return type and parameters are correctly set in the `symbol`. But in the `tSymbol` attribute which is a `TypeSymbol` those return type is sets as a `null` and parameters are sets to `0`.

Screenshot 2022-07-26 at 11 30 17

Screenshot 2022-07-26 at 11 30 58

In the above ss shows the debug results for the `'map` function. It has two parameters and the second parameter is a `BInvokableType` symbol. It has one parameter and the return type also set correctly. When accessing the Type Symbol those values are incorrectly created.

This issue causes for giving incorrect results `signature()` method for a `TypeSymbol`.

**Steps to reproduce:**

1. `LangLibFunctionTest -> testArrayLangLib()` run in debug mode.
2. `BIRPackageSymbolEnter -> defineFunction()`

**Suggested Labels (optional):**
Team/CompilerFE,Area/SemanticAPI,Priority/Blocker

### PR DESCRIPTION
**Pull request by chiranSachintha**
Fixes #36020
Fixes #37123

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Comment by MaryamZi** [comment]
@chiranSachintha, as discussed offline, let's add the tests for resource methods.

**Comment by MaryamZi** [comment]
Please check if the sonar bugs reported here are valid too. At least one of them can be resolved by extracting to variables.

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/values/ValueCreator.java [hunk 1]
```diff
@@ -90,6 +90,10 @@ public static ValueCreator getValueCreator(String key) {
         return runtimeValueCreators.get(key);
     }
 
+    public Object call(Strand strand, String funcName, Object... args) throws BError {
+        throw new BallerinaException();
```
Comment (warunalakshitha) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/values/ValueCreator.java:94]: Shall we retuen  something like new ErrorValue(StringUTils.fromString(No such method + fucntionnAme);

**Comment by chiranSachintha** [comment]
> SonarCloud Quality Gate failed.    [Quality Gate failed](https://sonarcloud.io/dashboard?id=ballerina-lang&pullRequest=37872)
> 
> [Bug](https://sonarcloud.io/project/issues?id=ballerina-lang&pullRequest=37872&resolved=false&types=BUG) [C](https://sonarcloud.io/project/issues?id=ballerina-lang&pullRequest=37872&resolved=false&types=BUG) [3 Bugs](https://sonarcloud.io/project/issues?id=ballerina-lang&pullRequest=37872&resolved=false&types=BUG) [Vulnerability](https://sonarcloud.io/project/issues?id=ballerina-lang&pullRequest=37872&resolved=false&types=VULNERABILITY) [A](https://sonarcloud.io/project/issues?id=ballerina-lang&pullRequest=37872&resolved=false&types=VULNERABILITY) [0 Vulnerabilities](https://sonarcloud.io/project/issues?id=ballerina-lang&pullRequest=37872&resolved=false&types=VULNERABILITY) [Security Hotspot](https://sonarcloud.io/project/security_hotspots?id=ballerina-lang&pullRequest=37872&resolved=false&types=SECURITY_HOTSPOT) [A](https://sonarcloud.io/project/security_hotspots?id=ballerina-lang&pullRequest=37872&resolved=false&types=SECURITY_HOTSPOT) [0 Security Hotspots](https://sonarcloud.io/project/security_hotspots?id=ballerina-lang&pullRequest=37872&resolved=false&types=SECURITY_HOTSPOT) [Code Smell](https://sonarcloud.io/project/issues?id=ballerina-lang&pullRequest=37872&resolved=false&types=CODE_SMELL) [A](https://sonarcloud.io/project/issues?id=ballerina-lang&pullRequest=37872&resolved=false&types=CODE_SMELL) [73 Code Smells](https://sonarcloud.io/project/issues?id=ballerina-lang&pullRequest=37872&resolved=false&types=CODE_SMELL)
> 
> [](https://sonarcloud.io/component_measures?id=ballerina-lang&pullRequest=37872&metric=new_coverage&view=list) [0.0% Coverage](https://sonarcloud.io/component_measures?id=ballerina-lang&pullRequest=37872&metric=new_coverage&view=list) [](https://sonarcloud.io/component_measures?id=ballerina-lang&pullRequest=37872&metric=new_duplicated_lines_density&view=list) [21.1% Duplication](https://sonarcloud.io/component_measures?id=ballerina-lang&pullRequest=37872&metric=new_duplicated_lines_density&view=list)

Will address this with a separate PR.

**Review by gimantha — APPROVED** [review]
LGTM

**Pull request merged**